### ISSUE TITLE
False positive in used_underscore_binding on `_next` in a for-loop

### ISSUE THREAD

**Issue by dtolnay** [Manishearth/rust-clippy#1846]
```rust
#![cfg_attr(feature = "cargo-clippy", deny(used_underscore_binding))]

pub fn f(bytes: &[u8]) {
    for b in bytes {
        println!("{}", b);
    }
}
```

```
error: used binding `_next` which is prefixed with an underscore. A leading underscore signals that a binding will not be used.
 --> src/lib.rs:4:5
  |
4 | /     for b in bytes {
5 | |         println!("{}", b);
6 | |     }
  | |_____^
  |
note: lint level defined here
 --> src/lib.rs:1:44
  |
1 | #![cfg_attr(feature = "cargo-clippy", deny(used_underscore_binding))]
  |                                            ^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#used_underscore_binding
```

**Comment by llogiq** [comment]
Interesting. They must've changed how a for loop is represented.

**Comment by dtolnay** [comment]
Ah, looks like https://github.com/rust-lang/rust/pull/42634.

**Comment by oli-obk** [comment]
This would have been fixed by https://github.com/rust-lang/rust/pull/39766, but unfortunately I can't really tell why that PR was bad and thus delayed...

**Comment by dtolnay** [comment]
Fixed by https://github.com/rust-lang/rust/pull/42871 in rustc 1.20.0-nightly (f590a44ce 2017-06-27). Thanks!

**Comment by oli-obk** [comment]
I'm fixing the for loop repr problems... All our `for` lints have stopped doing anything meaningful
